### PR TITLE
Execute complete type value

### PR DESCRIPTION
**Pull request by hungneox**
+ Bunch of function to complete different types
+ Add one important test

### CODE REVIEW

File: src/Execution/ExecutionStrategy.php [hunk 1]
```diff
@@ -85,20 +100,25 @@ protected function collectFields(
         $visitedFragmentNames
     ) {
         foreach ($selectionSet->getSelections() as $selection) {
-            /** @var FieldNode $selection */
             switch ($selection->getKind()) {
                 case NodeKindEnum::FIELD:
```
Comment (crisu83) [inline src/Execution/ExecutionStrategy.php]: Can't we use `instanceof` here instead of switch-case?

File: src/Execution/ExecutionStrategy.php [hunk 2]
```diff
@@ -127,6 +147,60 @@ protected function collectFields(
...
 
+
+    /**
+     * @param $node
+     * @return bool
+     * @throws InvalidTypeException
+     * @throws \Digia\GraphQL\Error\ExecutionException
+     * @throws \Digia\GraphQL\Error\InvariantException
+     */
+    private function shouldIncludeNode(NodeInterface $node)
```
Comment (Jalle19) [inline src/Execution/ExecutionStrategy.php]: : bool

File: src/Execution/ExecutionStrategy.php [hunk 3]
```diff
@@ -127,6 +147,60 @@ protected function collectFields(
...
+        return true;
+    }
+
+    /**
+     * @param FragmentDefinitionNode|InlineFragmentNode $fragment
+     * @param ObjectType                                $type
+     * @return bool
+     * @throws InvalidTypeException
+     */
+    private function doesFragmentConditionMatch(NodeInterface $fragment, ObjectType $type)
```
Comment (Jalle19) [inline src/Execution/ExecutionStrategy.php]: : bool

File: src/Execution/ExecutionStrategy.php [hunk 4]
```diff
@@ -138,19 +212,20 @@ private function getFieldNameKey(FieldNode $node)
...
+     * @param ObjectType $objectType
+     * @param            $rootValue
      * @param            $path
      * @param            $fields
-     *
      * @return array
      * @throws InvalidTypeException
+     * @throws \Digia\GraphQL\Error\ExecutionException
+     * @throws \Digia\GraphQL\Error\InvariantException
+     * @throws \Throwable
```
Comment (Jalle19) [inline src/Execution/ExecutionStrategy.php]: Where does \Throwable come from?

File: src/Execution/ExecutionStrategy.php [hunk 5]
```diff
@@ -284,58 +396,315 @@ private function buildResolveInfo(
...
+     */
+    private function completeValue(
+        TypeInterface $returnType,
+        $fieldNodes,
+        ResolveInfo $info,
+        $path,
+        &$result
+    ) {
+        if ($result instanceof \Exception || $result instanceof \Throwable) {
```
Comment (Jalle19) [inline src/Execution/ExecutionStrategy.php]: The first check is not needed, exceptions extend throwable

File: src/Execution/ExecutionStrategy.php (hunk 4 above)
Comment (hungneox) [reply src/Execution/ExecutionStrategy.php]: screen shot 2018-03-15 at 1 06 15
It's deep down somewhere, PhpStorm suggested to add that tag.

**Pull request merged**